x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1641,7 +1647,48 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
                 sslConfiguration = null;
                 populateClientCertificate = (channel, threadContext) -> {};
             }
-            return new Netty4HttpServerTransport(
+            final AuthenticationService authenticationService = this.authcService.get();
+            final ThreadContext threadContext = this.threadContext.get();
+            final TriConsumer<HttpPreRequest, Channel, ActionListener<ValidationResult>> authenticateMessage = (
+                httpRequest,
+                channel,
+                listener) -> {
+                // step 1: Wrap the passed-in listener such that the changes wrt to authentication are not visible to `listener#onResponse`.
+                // Authentication will change the thread-context, but we don't want such a context be visible to whatever code is going to
+                // run under the `listener#onResponse` call stack.
+                var contextPreservingListener = new ContextPreservingActionListener<>(
```

Comment:
I don't think this is necessary since we should not (nor should we) have anything that runs before this that depends on the thread context.

Reply:
I would like to answer this and https://github.com/elastic/elasticsearch/pull/95112#discussion_r1171634929 together.

After some further though, IMO, I think I prefer this function to not make assumptions upon the thread context.
This is code inside the Security plugin (where thread-context juggling happens the most), inside a lambda that is calling the `authenticationService.authenticate` that is known to alter the thread context, and this lambda, as designed herein protects the callers from that.
Therefore, IMO, the function is more robust to be calleable under any thread context and not alter it, as any normal function would.
I can also see your point that this function is callable only under very specific circumstances (because it refers to `perRequestThreadContext` which itself depends on the `httpPreRequest` parameter), so making it callable in "any" circumstances wrt to the thread-context is not really valuable.

If you still maintain that this lambda function should only work in the "default" thread context (the thread context of a netty transport worker), my follow-up question is: should we `assert` or throw exception if that expectation is not meet?